Define an exemplary format (#26)
ietf-wg-aipref/drafts: draft-ietf-aipref-vocab.md

The vocabulary does not preclude the use of other specific categories. Any statement of preference based on this vocabulary shall not be interpreted as restricting the use of the work(s) strictly for the purpose of search and discovery as long as no restriction is declared through search-specific means such as {{!RFC9309}}.

## Vocabulary Extensions {#vocab-extension}

Systems referencing the vocabulary must not introduce additional categories that include existing categories defined in the vocabulary or otherwise include additional hierarchical relationships.

> # Serialization Format {#format}

Reviewer: ```suggestion
# Exemplary Serialization Format {#format}
```

This section defines an exemplary serialization format for preferences.

> The format describes how the abstract model could be turned into a Unicode string or sequence of bytes.

Reviewer: By Unicode, you really mean UTF-8, yes?  If so, then say that since Unicode can be encoded in many ways.

Author: Hm, Unicode string has a definition that doesn't match what I wanted: https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-3/#G32765

What I'm looking for is probably Unicode text: a term without apparent definition, though Unicode uses it often.  Or maybe "coded character sequence": https://www.unicode.org/versions/Unicode16.0.0/core-spec/chapter-3/#G45138  (which is spectacularly opaque, so I might just use "Unicode text").

WDYT?

Author: ```suggestion
The format describes how the abstract model could be turned into Unicode text or sequence of bytes.
```

The format relies on the Dictionary type defined in {{Section 3.2 of !FIELDS=RFC9651}}.
The dictionary keys correspond to usage categories

> and the dictionary values correspond to explicit preferences.

Author: ```suggestion
and the dictionary values correspond to explicit preferences,
which can be either "y" or "n"; see {{y-or-n}}.
```

> 

Reviewer: Add something here about the values, `y` and `n`, so that it is clear what they represent.  Or just refer to the later section

For example, the following is a preference to allow AI training ({{ai}}),
disallow generative AI training ({{genai}}), and
and state no preference for other categories other than subsets of these categories:

~~~
ai=y, genai=n
~~~

## Usage Category Labels

Each usage category in the vocabulary ({{vocab}}) is mapped to a short textual label.
{{t-category-labels}} tabulates this mapping.

| Category               | Label | Reference |
|:-----------------------|:------|:----------|
| Text and Data Mining   | tdm   | {{tdm}}   |
| AI Training            | ai    | {{ai}}    |
| Generative AI Training | genai | {{genai}} |
{: #t-category-labels title="Mappings for Categories"}

Any mapping for a new usage category can only use

> lowercase latin characters (a-z), digits (0-9), "_", "-", ".", or "*".

Reviewer: I'd remove `*`, it has confusing semantics.  

Also, `.` is frequently used to separate "namespaces" - do we want to have that semantic?  If so, should we make that official?

Reviewer: I'll re-read 9651 and see if there are any issues with subsetting.

> These are encoded using the mappings in {{!ASCII=RFC0020}}.

Reviewer: Why do you say Unicode earlier, but now limit to ASCII?  Which is it?

Author: Given the range from which the codepoints are drawn, does it matter?

Reviewer: Technically no, but it's reads strange to me - which is why I added the comment.

> ## Preference Labels

Author: ```suggestion
## Preference Labels {#y-or-n}
```

> The abstract model used has two options for preferences associated with each category:

Reviewer: at this time - unless we add a third in the vocab.

And if we do, how would that be serialized?

allow and disallow.
These are mapped to single byte Tokens ({{Section 3.3.4 of !FIELDS}})

> of "y" and "n", respectively (without the quotes).

Reviewer: ```suggestion
of `y` and `n`, respectively.
```

Some formats are defined in terms of strings rather than bytes.
These formats might need to decode the bytes of this format to obtain a string.
As the syntax is limited to ASCII {{?ASCII=RFC0020}},
an ASCII decoder or UTF-8 decoder {{?UTF8=RFC3629}} can be used.
This results in the strings that this document uses.

Processing (see {{processing}}) requires a sequence of bytes,
so any format that uses strings needs to encode strings first.
Again, this process can use ASCII or UTF-8.

## Syntax Extensions

There are two ways by which this syntax might be extended:
the addition of new labels and the addition of parameters.

New labels might be defined to correspond to new usage categories.
{{vocab-extension}} addresses the considerations for defining new categories.
New labels might also be defined for other types of extension
that do not assign a preference to a usage category.
In either case, when processing a parsed Dictionary to obtain preferences,
any unknown labels MUST be ignored.

The Dictionary syntax ({{Section 3.2 of !FIELDS}}) can associate parameters
with each key-value pair.
This document does not define any semantics for any parameters that might be included.

> When processing a parsed Dictionary to obtain preferences,

Reviewer: So you think we should support parameters in the syntax but not provide any guidance on their usage?  Why not just disallow them?  Do we have a use case for them?

any unknown parameters MUST be ignored.

> ## Processing Algorithm {#processing}

Reviewer: Why do we need this here?  Isn't this defined in RFC 9651?  If so, just point there.

Author: The processing in RFC 9651 gets you a dictionary, not our abstract model.

To process a series of bytes to recover the expressed preferences,
those bytes are parsed into a Dictionary ({{Section 4.2.2 of !FIELDS}}),
then preferences are assigned to each usage category in the vocabulary.

The parsing algorithm for a Dictionary
produces a keyed collection of values,
each with a possibly-empty set of parameters.
The parsing process guarantees that each key has at most one value and parameters.

To obtain preferences for each of the categories in the vocabulary,
iterate through the categories.
For the label that corresponds to that category (see {{t-category-labels}}),
obtain the corresponding value from the collection,

> disregarding any parameters.

Reviewer: See - you are ignoring them anyway, so let's disallow from syntax to simplify the parsing rules!

A preference is assigned as follows:

> * If the value is a Token with a value of "y" (not including quotes),

Reviewer: ```suggestion
* If the value is a Token with a value of `y`,
```

  the associated preference is to allow that category of use.

> * If the value is a Token with a value of "n" (not including quotes),

Reviewer: ```suggestion
* If the value is a Token with a value of `n`,
```

  the associated preference is to disallow that category of use.

* Otherwise, a preference is not expressed for that category of use.

Note that this last alternative includes
the key being absent from the collection,
values that are not Tokens,
and Token values that are other than "y" or "n".
All of these are not errors,
they only result in no preference being inferred.

An important note about this process and format is that,
if the same key appears multiple times,
only the last value is taken.
This means that duplicating the same key could result in unexpected outcomes.
For example, the following expresses no preferences:

~~~
ai=y, ai="n", genai=n, genai, tdm=n, tdm=()
~~~

If the parsing of the Dictionary fails, no preferences are expressed.
This includes where keys include uppercase characters,

> as this format is case sensitive.

Reviewer: If this format is based on bytes and not strings - there is no such thing as case sensitivity, as that only applies to strings with a text encoding...which this does not have!

## Alternative Formats

> This format is only a recommended way to represent preferences.

Reviewer: ```suggestion
This format is only a exemplary way to represent preferences.
```

If the parsing of the Dictionary fails, no preferences are expressed.
This includes where keys include uppercase characters,
as this format is case sensitive
(more correctly, it operates on bytes, not strings).

## Alternative Formats

> This format is only a exemplary way to represent preferences.

Reviewer: ```suggestion
This format is only an exemplary way to represent preferences.
```
